sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala

```diff
@@ -220,7 +220,51 @@ object DeduplicateRelations extends Rule[LogicalPlan] {
           if (attrMap.isEmpty) {
             planWithNewChildren
           } else {
-            planWithNewChildren.rewriteAttrs(attrMap)
+            def rewriteAttrsMatchWithSubPlan(
+                attrs: Seq[Attribute],
+                attrMap: AttributeMap[Attribute],
+                planOutput: Seq[Attribute]): Seq[Attribute] = {
+              val canRewriteAttrs = attrMap.filter(a => planOutput.contains(a._2))
+              attrs.map(attr => canRewriteAttrs.getOrElse(attr, attr))
+            }
+
+            def rewriteOrderMatchWithSubPlan(
+                attrs: Seq[SortOrder],
+                attrMap: AttributeMap[Attribute],
+                planOutput: Seq[Attribute]): Seq[SortOrder] = {
+              val canRewriteAttrs = attrMap.filter(a => planOutput.contains(a._2))
+              attrs.filter(_.child.isInstanceOf[Attribute]).map(attr => {
+                if (canRewriteAttrs.contains(attr.child.asInstanceOf[Attribute])) {
+                  attr.copy(child = canRewriteAttrs(attr.child.asInstanceOf[Attribute]))
+                } else {
+                  attr
+                }
+              })
+            }
+
+            planWithNewChildren match {
+              case c: CoGroup =>
+                // SPARK-43781: CoGroup is a special case, as it has different output attributes
+                // from its children. We need to update the output attributes of CoGroup manually.
+                val newLeftAttr = c.leftAttr.map(attr => attrMap.getOrElse(attr, attr))
```

Comment:
why the left-right attributes does not need to call `rewriteAttrsMatchWithSubPlan`?

Reply:
should the same apple to `newRightAttr`?

Reply:
and to left/right group?